internal/forms/forms.go

```diff
@@ -237,6 +246,182 @@ func (m *Manager) GetFormTemplateHandler(w http.ResponseWriter, r *http.Request)
+		Action:        "update",
+	}, nil
+}
+
+func (m *Manager) getLatestFormsInfo() (string, string, error) {
+	client := http.Client{Timeout: 10 * time.Second}
+	req, _ := http.NewRequest("GET", formsVersionInfoUrl, nil)
+	req.Header.Set("User-Agent", m.config.UserAgent)
+	versionResp, err := client.Do(req)
+	if err != nil || versionResp.StatusCode != http.StatusOK {
+		err := fmt.Errorf("can't fetch winlink forms version page: %v", err)
+		log.Print(err)
+		return "", "", err
+	}
+	bodyBytes, err := ioutil.ReadAll(versionResp.Body)
+	if err != nil {
+		err := fmt.Errorf("can't read winlink forms version page: %v", err)
+		log.Print(err)
+		return "", "", err
+	}
+	bodyString := string(bodyBytes)
+
+	// Scrape for the version and download link
+	versionRe := regexp.MustCompile(`Standard_Forms - Version (\d+\.\d+\.\d+(\.\d+)?)`)
+	downloadRe := regexp.MustCompile(`https:\/\/1drv.ms\/u\/([a-zA-Z0-9-_!]+)\?e=([a-zA-Z0-9-_]+)`)
+	versionMatches := versionRe.FindStringSubmatch(bodyString)
+	downloadMatches := downloadRe.FindStringSubmatch(bodyString)
+	if versionMatches == nil || len(versionMatches) < 2 || downloadMatches == nil || len(downloadMatches) < 3 {
+		err := errors.New("can't scrape the version info page, HTML structure may have changed")
+		log.Print(err)
+		return "", "", err
+	}
+	newestVersion := versionMatches[1]
+	docId := downloadMatches[1]
+	auth := downloadMatches[2]
+	downloadLink := "https://api.onedrive.com/v1.0/shares/" + docId + "/root/content?e=" + auth
+	return newestVersion, downloadLink, nil
+}
+
+func (m *Manager) downloadAndUnzipForms(downloadLink string) error {
+	log.Printf("Updating forms via %v", downloadLink)
+	client := http.Client{Timeout: 30 * time.Second}
+	req, _ := http.NewRequest("GET", downloadLink, nil)
+	req.Header.Set("User-Agent", m.config.UserAgent)
+	downloadResp, err := client.Do(req)
+	if err != nil {
+		err := fmt.Errorf("can't download update ZIP: %v", err)
+		log.Print(err)
+		return err
+	}
+	filename := "Standard_Forms.zip"
+	dispo := downloadResp.Header.Get("Content-Disposition")
+	if dispo != "" && strings.Contains(dispo, "filename=") {
+		fileRe := regexp.MustCompile(`filename="(.*)"`)
+		filename = fileRe.FindStringSubmatch(dispo)[1]
+	}
+	dir, _ := ioutil.TempDir("", "pat")
+	defer os.RemoveAll(dir)
+	zipBytes, _ := ioutil.ReadAll(downloadResp.Body)
+	zipFilePath := path.Join(dir, filename)
+	err = ioutil.WriteFile(zipFilePath, zipBytes, 0600)
+	if err != nil {
+		err := fmt.Errorf("can't write update ZIP: %v", err)
+		log.Print(err)
+		return err
+	}
+
+	unzipDir := m.config.FormsPath
+	err = unzip(zipFilePath, unzipDir)
+	if err != nil {
+		err := fmt.Errorf("can't unzip forms update: %v", err)
+		log.Print(err)
+		return err
+	}
+	return nil
+}
+
+func unzip(src, dest string) error {
+	// https://stackoverflow.com/a/24792688/587091
+	r, err := zip.OpenReader(src)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Close(); err != nil {
+			panic(err)
+		}
```

Comment:
This was code I copied from SO. I suspect it's because this is in a deferred function, so we can't return the error if the reader doesn't close cleanly. I'm not especially attached to this logic; changed to simply `defer r.Close()`